R/run_stress_test.R

```diff
@@ -36,6 +36,58 @@ run_stress_test <- function(asset_type,
                             shock_year = 2030,
                             term = 2,
                             company_exclusion = TRUE) {
+  warn_if_asset_type_is_longer_than_1(asset_type)
+  matched_asset_type <- match.arg(asset_type)
+
+  # Allow testing anything before the long computation that follows
+  if (is_dev_env()) {
+    return(invisible(matched_asset_type))
+  }
+
+  # FIXME: Avoid clutter from printed "messages" that the user can't suppress
+  x <- capture.output(
+    run_stress_test_impl(
+      asset_type = matched_asset_type,
+      lgd_senior_claims = lgd_senior_claims,
+      lgd_subordinated_claims = lgd_subordinated_claims,
+      terminal_value = terminal_value,
+      risk_free_rate = risk_free_rate,
+      discount_rate = discount_rate,
+      div_netprofit_prop_coef = div_netprofit_prop_coef,
+      shock_year = shock_year,
+      term = term,
+      company_exclusion = company_exclusion
+    )
+  )
+
+  invisible(asset_type)
+}
+
+warn_if_asset_type_is_longer_than_1 <- function(asset_type) {
+  if (length(asset_type) > 1L) {
+    warn(c(
+      "`asset_type` should be of length 1.",
+      x = glue("It is of length {length(asset_type)}: {toString(asset_type)}."),
+      i = glue("Using default: {asset_type[[1]]}."),
+      i = glue(
+        "Suppress this warning e.g. with: `run_stress_test(asset_type = 'bonds')`."
+      )
```

Comment:
The `message` argument to `rlang::abort()` and friends is super cool!
https://rlang.r-lib.org/reference/abort.html#arguments

Reply:
- I put a general check that all input args must be of length 1 here https://github.com/2DegreesInvesting/r2dii.climate.stress.test/blob/52b74a86abcfa084c269ee86eb1e8d852b148c34/R/validate.R#L18, just making sure your test extends this?
- asset type being of length >1 would have to result in an error, not a warning.
- I wonder if this function should be specific to asset_type. A function warn_if_arg_is_longer_than_1 could be recycled for other inputs that need to be of length 1.

Reply:
1. Thanks, I saw that -- it's great you're asserting all inputs. But I think `asset_type` deserves special treatment. Compared to the other arguments, the possible values of this `asset_type` are small set of strings -- not a range of potentially infinite values. It fits really well in the description of the design pattern I link above. It's not a hard rule, of course, just a pattern. 
 
Here is one example in the wild:

``` r
head(usethis::use_blank_slate, 4)
#>                                           
#> 1 function (scope = c("user", "project")) 
#> 2 {                                       
#> 3     scope <- match.arg(scope)           
#> 4     if (scope == "user") {
```

2. If we use that pattern, then `match.arg()` is clever enough to throw informative errors.

``` r
f <- function(x = c("a", "b", "c")) {
  match.arg(x)
}

f(c("a", "b"))
#> Error in match.arg(x): 'arg' must be of length 1

f(c("d"))
#> Error in match.arg(x): 'arg' should be one of "a", "b", "c"
```

I'm also okay with using the first option silently, but I conservatively this time I chose to be more loud:

``` r
f <- function(x = c("a", "b", "c")) {
  if (length(x) > 1L) rlang::warn(glue::glue("Using default: {x[[1]]}."))
  match.arg(x)
}

f()
#> Warning: Using default: a.
#> [1] "a"
```

3. Yeah, as is this function is in an awckward middle gound. It's specific to `asset_type` but its name doesn't communicate that. Which way I take it depends on what we want:
    * a. Add "asset_type" to the name.
    * b. Make it general so we can use it elsewhere.
    
 The issue with b. is we might not need it. So I would start with a. and go with b. only if we later need it.

--

Of course, as every PR this is for discussion and I'm happy to keep the simpler approach we already have. Regardless how we handle `asset_type`, I think this PR already helped me think about how to test `run_stress_test()`. So happy to close and reuse the lessons I learn elsewhere.

Reply:
1. Thanks, I saw that -- it's great you're asserting all inputs. But I think `asset_type` deserves special treatment. Compared to the other arguments, the possible values of this `asset_type` are small set of strings -- not a range of potentially infinite values. It fits really well in the description of the design pattern I link above. It's not a hard rule, of course, just a pattern. 
 
Here is one example in the wild:

``` r
head(usethis::use_blank_slate, 4)
#>                                           
#> 1 function (scope = c("user", "project")) 
#> 2 {                                       
#> 3     scope <- match.arg(scope)           
#> 4     if (scope == "user") {
```

2. If we use that pattern, then `match.arg()` is clever enough to throw informative errors.

``` r
f <- function(x = c("a", "b", "c")) {
  match.arg(x)
}

f(c("a", "b"))
#> Error in match.arg(x): 'arg' must be of length 1

f(c("d"))
#> Error in match.arg(x): 'arg' should be one of "a", "b", "c"
```

I'm also okay with using the first option silently, but I conservatively this time I chose to be more loud:

``` r
f <- function(x = c("a", "b", "c")) {
  if (length(x) > 1L) rlang::warn(glue::glue("Using default: {x[[1]]}."))
  match.arg(x)
}

f()
#> Warning: Using default: a.
#> [1] "a"
```

3. Yeah, as is this function is in an awckward middle gound. It's specific to `asset_type` but its name doesn't communicate that. Which way I take it depends on what we want:
    * a. Add "asset_type" to the name.
    * b. Make it general so we can use it elsewhere.
    
 The issue with b. is we might not need it. So I would start with a. and go with b. only if we later need it.

--

Of course, as every PR this is for discussion and I'm happy to keep the simpler approach we already have. Regardless how we handle `asset_type`, I think this PR already helped me think about how to test `run_stress_test()`. So happy to close and reuse the lessons I learn elsewhere.

Reply:
1 and 3 ok
2 _I'm also okay with using the first option silently, but I conservatively this time I chose to be more loud:_ Just to rule out we have no misunderstanding. The first option may not be chosen. In case asset_type is not of length 1 we need to throw an error.

Reply:
actually another thought on 1.
We should keep in mind that we write our code for another audience. Who choses to use usethis is probably a developer and used to reading function documentation, which we cannot assume from our users who will be more guided by intuition.
```
Also as opposed to head(usethis::use_blank_slate, 4)
#>                                           
#> 1 function (scope = c("user", "project")) 
#> 2 {                                       
#> 3     scope <- match.arg(scope)           
#> 4     if (scope == "user") {
```
in our case we have no useable default.